File: server/grpc_server.go
package main

import (
	"fmt"
	"io"
	"log"
	"net"
	"os"
	"path/filepath"

	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"

	pb "gwbench/proto"
)

const (
	port       = ":50051"
	uploadDir  = "/app/uploads"
	chunkSize  = 4 * 1024 * 1024 // 4MB chunks for better performance
)

type fileServer struct {
	pb.UnimplementedFileServiceServer
}

func (s *fileServer) UploadFile(stream pb.FileService_UploadFileServer) error {
	log.Println("UploadFile called")
	
	var filename string
	var file *os.File
	var totalBytes int64
	
	defer func() {
		if file != nil {
			file.Close()
		}
	}()
	
	for {
		chunk, err := stream.Recv()
		if err == io.EOF {
			break
		}
		if err != nil {
			return status.Errorf(codes.Internal, "Failed to receive chunk: %v", err)
		}
		
		// First chunk - create file
		if file == nil {
			filename = chunk.Filename
			filepath := filepath.Join(uploadDir, filename)
			
			file, err = os.Create(filepath)
			if err != nil {
				return status.Errorf(codes.Internal, "Failed to create file: %v", err)
			}
			
			log.Printf("Receiving file: %s\n", filename)
		}
		
		// Write chunk directly to disk
		n, err := file.Write(chunk.Data)
		if err != nil {
			return status.Errorf(codes.Internal, "Failed to write chunk: %v", err)
		}
		
		totalBytes += int64(n)
		
		// Progress logging
		if totalBytes%(10*1024*1024) == 0 {
			log.Printf("Received %.1f MB\n", float64(totalBytes)/(1024*1024))
		}
	}
	
	log.Printf("File saved: %s (%d bytes)\n", filename, totalBytes)
	
	return stream.SendAndClose(&pb.UploadResponse{
		Message:       fmt.Sprintf("File %s uploaded successfully", filename),
		BytesReceived: totalBytes,
	})
}

func (s *fileServer) DownloadFile(req *pb.DownloadRequest, stream pb.FileService_DownloadFileServer) error {
	log.Printf("DownloadFile called for: %s\n", req.Filename)
	
	filepath := filepath.Join(uploadDir, req.Filename)
	
	file, err := os.Open(filepath)
	if err != nil {
		return status.Errorf(codes.NotFound, "File %s not found", req.Filename)
	}
	defer file.Close()
	
	buffer := make([]byte, chunkSize)
	var bytesSent int64
	
	for {
		n, err := file.Read(buffer)
		if err == io.EOF {
			break
		}
		if err != nil {
			return status.Errorf(codes.Internal, "Failed to read file: %v", err)
		}
		
		chunk := &pb.FileChunk{
			Filename: req.Filename,
			Data:     buffer[:n],
		}
		
		if err := stream.Send(chunk); err != nil {
			return status.Errorf(codes.Internal, "Failed to send chunk: %v", err)
		}
		
		bytesSent += int64(n)
		
		// Progress logging
		if bytesSent%(10*1024*1024) == 0 {
			log.Printf("Sent %.1f MB\n", float64(bytesSent)/(1024*1024))
		}
	}
	
	log.Printf("Download complete: %d bytes sent\n", bytesSent)
	return nil
}

func main() {
	// Create upload directory
	if err := os.MkdirAll(uploadDir, 0755); err != nil {
		log.Fatalf("Failed to create upload directory: %v", err)
	}
	
	log.Printf("FileService initialized with upload_dir: %s\n", uploadDir)
	
	lis, err := net.Listen("tcp", port)
	if err != nil {
		log.Fatalf("Failed to listen: %v", err)
	}
	
	// Create gRPC server with increased message size limits
	s := grpc.NewServer(
		grpc.MaxRecvMsgSize(500*1024*1024), // 500MB
		grpc.MaxSendMsgSize(500*1024*1024), // 500MB
	)
	
	pb.RegisterFileServiceServer(s, &fileServer{})
	
	log.Printf("Server listening on %s\n", port)
	
	if err := s.Serve(lis); err != nil {
		log.Fatalf("Failed to serve: %v", err)
	}
}

File: server/grpc_server.py
import grpc
from concurrent import futures
import file_service_pb2
import file_service_pb2_grpc
import os
import time
import sys

class FileService(file_service_pb2_grpc.FileServiceServicer):
    def __init__(self):
        self.upload_dir = "/app/uploads"
        os.makedirs(self.upload_dir, exist_ok=True)
        print(f"FileService initialized with upload_dir: {self.upload_dir}", flush=True)

    def UploadFile(self, request_iterator, context):
        print("UploadFile called", flush=True)
        filename = None
        total_bytes = 0
        file_data = b""
        
        for chunk in request_iterator:
            if filename is None:
                filename = chunk.filename
                print(f"Receiving file: {filename}", flush=True)
            file_data += chunk.data
            total_bytes += len(chunk.data)
            if total_bytes % (10 * 1024 * 1024) == 0:
                print(f"Received {total_bytes / (1024*1024):.1f} MB", flush=True)
        
        if filename:
            filepath = os.path.join(self.upload_dir, filename)
            with open(filepath, 'wb') as f:
                f.write(file_data)
            print(f"File saved: {filepath} ({total_bytes} bytes)", flush=True)
        
        return file_service_pb2.UploadResponse(
            message=f"File {filename} uploaded successfully",
            bytes_received=total_bytes
        )

    def DownloadFile(self, request, context):
        print(f"DownloadFile called for: {request.filename}", flush=True)
        filepath = os.path.join(self.upload_dir, request.filename)
        
        if not os.path.exists(filepath):
            context.set_code(grpc.StatusCode.NOT_FOUND)
            context.set_details(f"File {request.filename} not found")
            return
        
        chunk_size = 64 * 1024  # 64KB chunks
        bytes_sent = 0
        with open(filepath, 'rb') as f:
            while True:
                chunk_data = f.read(chunk_size)
                if not chunk_data:
                    break
                bytes_sent += len(chunk_data)
                if bytes_sent % (10 * 1024 * 1024) == 0:
                    print(f"Sent {bytes_sent / (1024*1024):.1f} MB", flush=True)
                yield file_service_pb2.FileChunk(
                    filename=request.filename,
                    data=chunk_data
                )
        print(f"Download complete: {bytes_sent} bytes sent", flush=True)

def serve():
    print("Starting gRPC server...", flush=True)
    server = grpc.server(
        futures.ThreadPoolExecutor(max_workers=10),
        options=[
            ('grpc.max_send_message_length', 500 * 1024 * 1024),
            ('grpc.max_receive_message_length', 500 * 1024 * 1024),
        ]
    )
    file_service_pb2_grpc.add_FileServiceServicer_to_server(FileService(), server)
    listen_addr = '[::]:50051'
    server.add_insecure_port(listen_addr)
    
    print(f"Server listening on {listen_addr}", flush=True)
    server.start()
    print("Server started successfully", flush=True)
    
    try:
        server.wait_for_termination()
    except KeyboardInterrupt:
        print("Server stopped", flush=True)
        server.stop(0)

if __name__ == '__main__':
    serve()